File: CIS194_intro_haskell/2013_spring/hw1/card.ml
let validate (cardNum : int) : bool =
  cardNum |> num2digits |> doubleOddPositions |> sumDigits
  |> fun x -> x mod 10 == 0


let num2digits (num : int) : int list =
  let rec _num2digits (n : int) (acc : int list) : int list =
    if n < 10 then n :: acc
    else
      let d = n / 10 in
      let r = n mod 10 in
      _num2digits d (r :: acc)
  in
  _num2digits num []


let doubleOddPositions (nums : int list) : int list =
  let f (acc : bool * int list) (num : int) : (bool * int list) =
    match acc with
    | (true, a) -> (false, (num * 2) :: a)
    | (false, a) -> (true, num :: a)
  in
  List.fold_left f (true, []) nums
  |> snd
  |> List.rev


let rec sumDigits (digits : int list) : int =
  let f (a : int) (d : int) : int =
    a + if d < 10 then d
        else d |> num2digits |> sumDigits
  in
  List.fold_left f 0 digits

File: CIS194_intro_haskell/2013_spring/hw1/card.scala
object Card {

  def validate(cardNum: Long): Boolean = {
    val code = sumDigits(doubleOddPositions(num2digits(cardNum)))
    code % 10 == 0
  }

  def num2digits(num: Long): List[Int] = {
    def _num2digits(num: Long): List[Int] = {
      if (num < 10) List(num.toInt)
      else {
        val d = num / 10
        val r = (num % 10).toInt
        r :: _num2digits(d)
      }
    }
    _num2digits(num).reverse
  }

  def doubleOddPositions(nums: List[Int]): List[Int] = {
    def _even(n: List[Int]): List[Int] = n match {
      case Nil => Nil
      case h :: t => (h * 2) :: _odd(t)
    }
    def _odd(n: List[Int]): List[Int] = n match {
      case Nil => Nil
      case h :: t => h :: _even(t)
    }
    _even(nums)
  }

  def sumDigits(digits: List[Int]): Int = {
    digits.foldLeft(0) {
      (acc, d) => {
        if(d < 10) d
        else sumDigits(num2digits(d))
      } + acc
    }
  }
}
